group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/TopologyMetadata.java

```diff
@@ -0,0 +1,115 @@
+    }
+
+    /**
+     * Map of topic names to topic metadata.
+     *
+     * @return The map of topic Ids to topic metadata.
+     */
+    @Override
+    public Map<String, TopicMetadata> topicMetadata() {
+        return this.topicMetadata;
+    }
+
+    /**
+     * Checks whether the given subtopology is associated with a changelog topic.
+     *
+     * @param subtopologyId String identifying the subtopology.
+     * @throws IllegalStateException if the subtopology ID does not exist, or the topology is not configured.
+     * @return true if the subtopology is associated with a changelog topic, false otherwise.
+     */
+    @Override
+    public boolean isStateful(String subtopologyId) {
+        final ConfiguredSubtopology subtopology = getSubtopologyOrFail(subtopologyId);
+        return !subtopology.stateChangelogTopics().isEmpty();
+    }
+
+    /**
+     * The list of subtopologies in the topology.
+     *
+     * @throws IllegalStateException if the topology is not configured.
+     * @return a list of subtopology IDs.
+     */
+    @Override
+    public List<String> subtopologies() {
+        return getSubtopologiesOrFail().keySet().stream().toList();
+    }
+
+    /**
+     * The maximal number of input partitions among all source topics for the given subtopology.
+     *
+     * @param subtopologyId String identifying the subtopology.
+     *
+     * @throws IllegalStateException if the subtopology ID does not exist, or the topology is not configured, or the subtopology
+     *                               contains no source topics.
+     * @return The maximal number of input partitions among all source topics for the given subtopology.
+     */
+    @Override
+    public int maxNumInputPartitions(String subtopologyId) {
+        final ConfiguredSubtopology subtopology = getSubtopologyOrFail(subtopologyId);
+        return Stream.concat(
+            subtopology.sourceTopics().stream(),
+            subtopology.repartitionSourceTopics().keySet().stream()
+        ).map(topic -> this.topicMetadata.get(topic).numPartitions()).max(Integer::compareTo).orElseThrow(
+            () -> new IllegalStateException("Subtopology does not contain any source topics")
+        );
+    }
+
+    private ConfiguredSubtopology getSubtopologyOrFail(String subtopologyId) {
+        final Map<String, ConfiguredSubtopology> subtopologies = getSubtopologiesOrFail();
+        if (!subtopologies.containsKey(subtopologyId)) {
+            throw new IllegalStateException(String.format("Topology does not contain subtopology %s", subtopologyId));
+        }
+        return subtopologies.get(subtopologyId);
+    }
+
+    private Map<String, ConfiguredSubtopology> getSubtopologiesOrFail() {
+        final Optional<SortedMap<String, ConfiguredSubtopology>> subtopologies = topology.subtopologies();
+        if (subtopologies.isEmpty()) {
+            throw new IllegalStateException("Topology is not configured");
+        }
```

Comment:
Is there a reason, you cannot make this check at construction time?

Reply:
After some offline discussion, we agreed to unpack the optional inside `TopologyMetadata` to avoid the `isPresent` check altogether.